include/deal.II/lac/linear_operator.h

```diff
@@ -2126,6 +2126,289 @@ operator*(const PackagedOperation<Range> &comp,
+      {
+        v.block(i) *= 0;
+        for (unsigned int j = i + 1; j < a.n_block_cols(); ++j)
+          a.block(i,j).Tvmult_add(v.block(i), u.block(j));
+      }
+
+  };
+
+  return_op.vmult_add = [&a](Range &v, const Domain &u)
+  {
+    Assert( v.n_blocks() == a.n_block_rows(),
+            ExcDimensionMismatch(v.n_blocks(), a.n_block_rows()));
+    Assert( u.n_blocks() == a.n_block_cols(),
+            ExcDimensionMismatch(u.n_blocks(), a.n_block_cols()));
+
+    for (unsigned int i = 0; i < a.n_block_cols(); ++i)
+      for (unsigned int j = i + 1; j < a.n_block_rows(); ++j)
+        a.block(i,j).Tvmult_add(v.block(i), u.block(j));
+  };
+
+  return_op.Tvmult = [&a](Domain &v, const Range &u)
+  {
+    Assert( v.n_blocks() == a.n_block_cols(),
+            ExcDimensionMismatch(v.n_blocks(), a.n_block_cols()));
+    Assert( u.n_blocks() == a.n_block_rows(),
+            ExcDimensionMismatch(u.n_blocks(), a.n_block_rows()));
+
+    for (unsigned int i = 0; i < a.n_block_cols(); ++i)
+      {
+        v.block(i) *= 0;
+        for (unsigned int j = 0; j < i; ++j)
+          a.block(j,i).vmult_add(v.block(i), u.block(j));
+      }
+  };
+
+  return_op.Tvmult_add = [&a](Domain &v, const Range &u)
+  {
+    Assert( v.n_blocks() == a.n_block_cols(),
+            ExcDimensionMismatch(v.n_blocks(), a.n_block_cols()));
+    Assert( u.n_blocks() == a.n_block_rows(),
+            ExcDimensionMismatch(u.n_blocks(), a.n_block_rows()));
+
+    for (unsigned int i = 0; i < a.n_block_cols(); ++i)
+      for (unsigned int j = 0; j < i; ++j)
+        a.block(j,i).vmult_add(v.block(i), u.block(j));
+  };
+
+  return return_op;
+}
+
+
+/**
+ * @relates LinearOperator
+ *
+ * Let M be a block matrix of the form Id + T made of nxn blocks
+ * and where T is a lower diagonal / upper diagonal (without diagonal).
+ * Then, its inverse is of the form:
+ * @code
+ *  Id + sum_{i=1}^{n-1} (-1)^i T^i
+ * @endcode
+ * This formula can be used to invert all triangular matrix
+ * (diagonal included of course).
+ *
+ * This function takes a block matrix @p a (possibly full) and a linear block operator
+ * @p inverse_diagonal made of the inverse of the diagonal blocks inverses.
+ * The output is the inverse of the matrix in the case of a triangular matrix and
+ * as inverse_diagonal its diagonal blocks inverses.
+ * Otherwise, the result is a preconditioner.
+ *
+ * The parameter @p lower is a bool that allows to specify if we want
+ * to use lower triangular part of a (true, this is the default value) or
+ * to use upper triangular part of a (false).
+ * @ingroup LAOperators
+ */
+template <typename Range = BlockVector<double>,
+          typename Domain = Range,
+          typename BlockMatrix>
+LinearOperator<Range, Domain>
+preconditioner_from_diagonal_inverse( BlockMatrix &a,
+                                      const LinearOperator<Range, Domain> &inverse_diagonal,
+                                      bool lower = true)
```

Comment:
make `lower` `const` to indicate that it is an input argument. Could you make `a` const as well? (Optional: using little-a as the name for a matrix may be a bit confusing. Maybe uppercase-A? or `block_matrix`?)